plugins/steelseries/fu-steelseries-sonic.c

```diff
@@ -0,0 +1,1131 @@
+					     firmware,
+					     fu_progress_get_child(progress),
+					     flags,
+					     error))
+		return FALSE;
+	fu_progress_step_done(progress);
+	if (!fu_steelseries_sonic_verify_chip(device,
+					      chip,
+					      firmware,
+					      fu_progress_get_child(progress),
+					      error))
+		return FALSE;
+	fu_progress_step_done(progress);
+
+	/* holtek */
+	chip = STEELSERIES_SONIC_CHIP_HOLTEK;
+	if (!fu_steelseries_sonic_write_chip(device,
+					     STEELSERIES_SONIC_CHIP_HOLTEK,
+					     firmware,
+					     fu_progress_get_child(progress),
+					     flags,
+					     error))
+		return FALSE;
+	fu_progress_step_done(progress);
+	if (!fu_steelseries_sonic_verify_chip(device,
+					      chip,
+					      firmware,
+					      fu_progress_get_child(progress),
+					      error))
+		return FALSE;
+	fu_progress_step_done(progress);
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_steelseries_sonic_parse_firmware(FuFirmware *firmware, FwupdInstallFlags flags, GError **error)
+{
+	guint32 checksum_tmp;
+	guint32 checksum;
+	g_autoptr(GBytes) blob = NULL;
+
+	blob = fu_firmware_get_bytes(firmware, error);
+	if (blob == NULL)
+		return FALSE;
+
+	if (!fu_common_read_uint32_safe(g_bytes_get_data(blob, NULL),
+					g_bytes_get_size(blob),
+					g_bytes_get_size(blob) - sizeof(checksum),
+					&checksum,
+					G_LITTLE_ENDIAN,
+					error))
+		return FALSE;
+	checksum_tmp = fu_common_crc32(g_bytes_get_data(blob, NULL),
+				       g_bytes_get_size(blob) - sizeof(checksum_tmp));
+	checksum_tmp = ~checksum_tmp;
+	if (checksum_tmp != checksum) {
+		if ((flags & FWUPD_INSTALL_FLAG_IGNORE_CHECKSUM) == 0) {
+			g_set_error(error,
+				    FWUPD_ERROR,
+				    FWUPD_ERROR_INTERNAL,
+				    "checksum mismatch for %s, got 0x%08x, expected 0x%08x",
+				    fu_firmware_get_id(firmware),
+				    checksum_tmp,
+				    checksum);
+			return FALSE;
+		}
+		g_debug("ignoring checksum mismatch, got 0x%08x, expected 0x%08x",
+			checksum_tmp,
+			checksum);
+	}
+
+	fu_firmware_add_flag(firmware, FU_FIRMWARE_FLAG_HAS_CHECKSUM);
+
+	/* success */
+	return TRUE;
+}
+
+static FuFirmware *
+fu_steelseries_sonic_prepare_firmware(FuDevice *device,
+				      GBytes *fw,
+				      FwupdInstallFlags flags,
+				      GError **error)
+{
+	SteelseriesSonicChip chip;
+	g_autoptr(FuFirmware) firmware_nordic = NULL;
+	g_autoptr(FuFirmware) firmware_holtek = NULL;
+	g_autoptr(FuFirmware) firmware_mouse = NULL;
+	g_autoptr(FuFirmware) firmware_chip = NULL;
+	g_autoptr(FuFirmware) firmware = NULL;
+
+	firmware = fu_archive_firmware_new();
+	if (!fu_firmware_parse(firmware, fw, flags, error))
+		return FALSE;
+
+	/* mouse */
+	chip = STEELSERIES_SONIC_CHIP_MOUSE;
+	firmware_chip =
+	    fu_firmware_get_image_by_id(firmware, STEELSERIES_SONIC_FIRMWARE_ID[chip], error);
+	if (firmware_chip == NULL)
+		return NULL;
+	if (!fu_steelseries_sonic_parse_firmware(firmware_chip, flags, error))
+		return FALSE;
+	firmware_mouse = g_steal_pointer(&firmware_chip);
```

Comment:
i'm unclear what you need firmware_mouse for... can we maybe do:

	g_autoptr(FuFirmware) firmware_chip1 = NULL;
	firmware_chip1 = fu_firmware_get_image_by_id(firmware, STEELSERIES_SONIC_FIRMWARE_ID[chip], error);
	if (firmware_chip1 == NULL)
		return NULL;

etc?

Reply:
Well, the idea have the exact same code for the three bloc of code.

I wanted to have:

``` c
g_autoptr(FuFirmware) firmwares[_LAST] = { NULL };

firmwares[chip] = fu_firmware_get_image_by_id(firmware, STEELSERIES_SONIC_FIRMWARE_ID[chip], error);
if (firmwares[chip] == NULL)
	return NULL;
```

I do no know if this is doable.